### PR TITLE
Type hints MCTCT

### PR DESCRIPTION
**Pull request by rchan26**
Hi @Rocketknight1: this PR looks to add type hints to MCTCT models and addresses #16059.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/models/mctct/modeling_mctct.py [hunk 1]
```diff
@@ -751,14 +751,14 @@ def __init__(self, config):
...
-        labels=None,
-    ):
+        input_features: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        labels: Optional[torch.LongTensor] = None,
+    ) -> Union[Tuple, BaseModelOutput]:
```
Comment (Rocketknight1) [inline src/transformers/models/mctct/modeling_mctct.py:761]:
```suggestion
    ) -> Union[Tuple, CausalLMOutput]:
```

File: src/transformers/models/mctct/modeling_mctct.py [hunk 2]
```diff
@@ -566,13 +566,13 @@ def __init__(self, config: MCTCTConfig):
...
-        input_features,
-        attention_mask,
-        head_mask,
-        output_attentions=False,
-        output_hidden_states=False,
-        return_dict=True,
-    ):
+        input_features: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/mctct/modeling_mctct.py:571]: Careful, these weren't optional before!

**Review by Rocketknight1 — APPROVED** [review]
This looks mostly good, thanks! One thing to watch out for though is that in the encoder `forward()` method, `attention_mask` and `head_mask` weren't optional before - make sure you don't change any default values when setting the type hints!

If you fix up that detail and change the return type for the CTC model this looks ready to go, though!

File: src/transformers/models/mctct/modeling_mctct.py (hunk 2 above)
Comment (rchan26) [reply src/transformers/models/mctct/modeling_mctct.py]: I see - I've removed the `Optional` parts for these variables now

**Comment by rchan26** [comment]
So it seems like removing the `Optional` has introduced an error in the tests - looks like the test doesn't supply anything for `head_mask`?

**Comment by Rocketknight1** [comment]
Ah, careful! `attention_mask` and `head_mask` **were** optional in the main model methods, but not in the encoder. The encoder mostly doesn't have `Optional` arguments, but the other classes do. The changes to the main model methods is why you're getting that error now!

This is much easier to see if you look at the 'files changed' interface on GitHub - we want this PR to only add type annotations, but not change any default arguments.

**Comment by rchan26** [comment]
I see - my apologies! Will make some changes now to fix this :)

File: src/transformers/models/mctct/modeling_mctct.py [hunk 3]
```diff
@@ -566,13 +566,13 @@ def __init__(self, config: MCTCTConfig):
...
-        output_attentions=False,
-        output_hidden_states=False,
-        return_dict=True,
-    ):
+        input_features: torch.Tensor,
+        attention_mask: torch.Tensor,
+        head_mask: torch.Tensor,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
```
Comment (Rocketknight1) [inline src/transformers/models/mctct/modeling_mctct.py:574]:
```suggestion
        output_attentions: bool = False,
        output_hidden_states: bool = False,
        return_dict: bool = True,
```

**Review by Rocketknight1 — APPROVED** [review]
Looks good now! I made one small fix to the encoder but it's otherwise perfect.

**Pull request merged**